sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -44,82 +44,116 @@ RingOutlierFilterComponent::RingOutlierFilterComponent(const rclcpp::NodeOptions
-  for (const auto & ring_indices : input_ring_map) {
-    if (ring_indices.second.size() < 2) {
-      continue;
-    }
 
-    for (size_t idx = 0U; idx < ring_indices.second.size() - 1; ++idx) {
-      const auto & current_idx = ring_indices.second.at(idx);
-      const auto & next_idx = ring_indices.second.at(idx + 1);
-      tmp_indices.emplace_back(current_idx);
+  std::vector<std::vector<size_t>> ring2indices;
+  ring2indices.reserve(128);
+
+  for (size_t i = 0; i < 128; i++) {
+    ring2indices.push_back(std::vector<size_t>());
+    ring2indices.back().reserve(2000);  // TODO(sykwer): Make it parameter
+  }
+
+  for (size_t data_idx = 0; data_idx < input->data.size(); data_idx += input->point_step) {
+    const uint16_t ring = *reinterpret_cast<const uint16_t *>(&input->data[data_idx + ring_offset]);
+    ring2indices[ring].push_back(data_idx);
+  }
+
+  // walk range: [walk_first_idx, walk_last_idx]
+  int walk_first_idx = 0;
+  int walk_last_idx = -1;
+
+  for (const auto & indices : ring2indices) {
+    if (indices.size() < 2) continue;
+
+    walk_first_idx = 0;
+    walk_last_idx = -1;
+
+    for (size_t idx = 0U; idx < indices.size() - 1; ++idx) {
+      const size_t & current_data_idx = indices[idx];
+      const size_t & next_data_idx = indices[idx + 1];
+      walk_last_idx = idx;
 
       // if(std::abs(iter->distance - (iter+1)->distance) <= std::sqrt(iter->distance) * 0.08)
-      const auto current_pt_azimuth =
-        *reinterpret_cast<float *>(&input_ptr->data[current_idx + azimuth_offset]);
-      const auto next_pt_azimuth =
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + azimuth_offset]);
-      float azimuth_diff = next_pt_azimuth - current_pt_azimuth;
+
+      const float & current_azimuth =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + azimuth_offset]);
+      const float & next_azimuth =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + azimuth_offset]);
+      float azimuth_diff = next_azimuth - current_azimuth;
       azimuth_diff = azimuth_diff < 0.f ? azimuth_diff + 36000.f : azimuth_diff;
 
-      const auto current_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[current_idx + distance_offset]);
-      const auto next_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + distance_offset]);
+      const float & current_distance =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + distance_offset]);
+      const float & next_distance =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + distance_offset]);
 
       if (
-        std::max(current_pt_distance, next_pt_distance) <
-          std::min(current_pt_distance, next_pt_distance) * distance_ratio_ &&
+        std::max(current_distance, next_distance) <
+          std::min(current_distance, next_distance) * distance_ratio_ &&
         azimuth_diff < 100.f) {
```

Comment:
should `azimuth_diff < 100.f` be a parameter?

Reply:
This number has been hard-coded for a long period of time and should, of course, be parameterised, but this is not necessary.